### PR TITLE
MSOutput: stop output placement for relval workflows by default

### LINKED ISSUE #12069: Stop relval output placement in MSOutput in favor of rucio subscriptions
**Impact of the new feature**
MSOutput

**Is your feature request related to a problem? Please describe.**
Currently relval output placement is done via MSOutput and changing the algorithm requires code changes which takes developer time. We'd like to change the relval output policy and start using Rucio subscription mechanism for relval placement instead of MSOutput, which gives CompOps the full control to change the policy as needed.

**Describe the solution you'd like**
Stop output placement for relval workflows. @amaltaro please let us know the date you'll deploy this change, so that we can start the relevant rucio subscriptions a few days in advance. After we make sure that subscriptions work, we can disable MSOutput relval logic.

**Describe alternatives you've considered**
None

**Additional context**
@Panos512 @drkovalskyi fyi

If anyone is interested in the new policy, please let me know, I can send it via private chat.

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #12069

#### Status
 not-tested

#### Description
As requested by P&R, add configuration to stop output placement for relval workflows

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Documentation: https://gitlab.cern.ch/dmwm/wmcore-docs/-/merge_requests/50/diffs

#### External dependencies / deployment changes
Service configuration changes:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/307
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/308
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/309

### CODE REVIEW

**Comment by khurtado** [comment]
@amaltaro This should skip relval workflow (they would not be uploaded to MongoDB, hence not processed by the MSOutput consumer).

I made it configurable with `True` as default. Would this configuration be enough?
Can we use testbed MSOutput to test?

**Comment by amaltaro** [comment]
@khurtado Kenyi, I feel like we should keep producing output documents (persisted in MongoDB) for RelVal, but just don't perform any actual Rucio rule creation if configured so.

About the service configuration, perhaps we could make it more inline with the Tape one:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/blob/prod/reqmgr2ms-output/config-output.py#L78
and call it something like `enableRelValDisk`.

Note that such RelVal documents would still have to be marked as `TransferStatus=done`, by MSOutput. Here is one example MSOutput document, despite it having actual rucio rules:
https://cmsweb.cern.ch:8443/ms-output/data/info?request=cmsunified_task_BPH-RunIISummer20UL16GEN-00197__v1_T_231129_095358_1095

**Comment by khurtado** [comment]
@amaltaro Ok, I changed things in a way rucio disk/tape subscriptions are skipped but transfers are marked as done in the MongoDB workflow. Was this what you meant?

File: src/python/WMCore/MicroService/MSOutput/MSOutput.py [hunk 1]
```diff
@@ -310,6 +313,15 @@ def makeSubscriptions(self, workflow):
         # add a configurable weight value
         ruleAttrs["weight"] = self.msConfig['rucioDiskRuleWeight']
 
+        # Do not make subscription for relval workflow if not enabled
+        # and consider transfer as done
+        if workflow['IsRelVal'] and self.msConfig['enableRelValDisk']:
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSOutput/MSOutput.py:318]: `and not self.msConfig['enableRelValDisk']:`?

I would suggest not to use the word "subscription" in the message below as well, as subscriptions are different between the old PhEDEx and Rucio. Perhaps something like "Skipping disk output data placement..."

File: src/python/WMCore/MicroService/MSOutput/MSOutput.py [hunk 2]
```diff
@@ -310,6 +313,15 @@ def makeSubscriptions(self, workflow):
...
+            self.logger.info("Skipping disk subscription for relval workflow",
+                             workflow['RequestName'])
+            workflow.updateTime()
+            self.docKeyUpdate(workflow, TransferStatus='done')
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSOutput/MSOutput.py:322]: I think we do not need to have these lines of code over here.
We can likely just let the workflow go through the Pipeline and it should naturally be marked as `done`, of course evaluating whether or not disk data placement needs to be performed or not.

For instance, the code above would likely kick in at:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSOutput/MSOutput.py#L362

File: src/python/WMCore/MicroService/MSOutput/MSOutput.py [hunk 3]
```diff
@@ -382,6 +394,12 @@ def makeTapeSubscriptions(self, workflow):
         :param workflow: a MSOutputTemplate object representing a workflow
         :return: the MSOutputTemplate object itself (with the necessary updates in place)
         """
+        # Do not make relval workflow subscription if not enabled
+        if workflow['IsRelVal'] and self.msConfig['enableRelValDisk']:
+            self.logger.info("Skipping tape subscription for relval workflow",
+                             workflow['RequestName'])
+            return workflow
```
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSOutput/MSOutput.py:401]: Similarly to the comment above, I fear that if we simply return here, we might miss the execution of some logic in this function of the pipeline.
Perhaps another way to implement it more inline with the current implementation of this microservice would be to check for relval disk data placement right at the beginning of this for loop:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSOutput/MSOutput.py#L394
`continue`ing to the next loop iteration if necessary.

**Review by amaltaro — CHANGES_REQUESTED** [review]
@khurtado the logic implemented in correct, but I feel like it skips the actual pipeline implemented in each of the modes implemented for this microservice (producer/consumer). I left a few comments along the code.

You might want to refer to this documentation for further background information:
https://gitlab.cern.ch/dmwm/wmcore-docs/-/blob/master/docs/microservices/ReqMgr2-MicroService-Output.md?ref_type=heads

which we will have to make minor updates as well, as we move forward with these changes. Please update the documentation accordingly as well.

**Comment by khurtado** [comment]
@amaltaro Thanks for the feedback! I overlooked some methods in the pipeline. I simplified the code and made it more in line with other setting decisions in the pipeline.

**Comment by amaltaro** [comment]
Beautiful! Even better than what I initially thought, thanks!

@khurtado before approving this PR, can you please create the relevant services_config MRs (for test, preprod and prod branches) by explicitly defining this parameter and by default setting it to false?

In addition, please update the documentation with this new hook/detail.

**Comment by khurtado** [comment]
@amaltaro 

Here are the merge request changes for the services config on all 3 branches:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/307
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/308
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/309

And documentation:
https://gitlab.cern.ch/dmwm/wmcore-docs/-/merge_requests/50/diffs

**Review by amaltaro — APPROVED** [review]
Thanks Kenyi! These changes look good to me.
I have also copied the configuration and documentation links to the initial PR description as well.

**Pull request merged**